# Práctica 6, Grupo 02 (Mario Calvarro, Beñat Pérez)

## Representación cromosomas
Para representar cada cromosoma hemos optado por una lista de 900 listas de seis enteros cada una. Estos enteros representan la cantidad de minas de un determinado tipo (en orden) que se compran en ese segundo. 

## Decodificación
En nuestro caso el fenotipo es igual al genotipo.

## Población inicial
Para generar la población inicial hemos utilizado la siguiente asignación:

In [ ]:
number: int = int(random.randint(0, 1) * 1.045 **(i-90))
number = round(number * 3**(-6.5*ii))

Es decir, tiene un crecimiento exponencial respecto de los segundos e inversamente exponencial respecto de las minas. Por tanto, los últimos segundos se compraran muchas más minas y las minas de precio más bajo se compraran en mayor cantidad.

Los últimos 20 segundos no se compra ninguna mina.

## Operadores genéticos
### Selección
Para seleccionar que individuos se cruzaran o pasarán a la siguiente generación utilizamos una selección por torneo de grupos de 3 individuos donde tomamos aquel que tenga mayor fitness (oro al final de la partida).

### Mutación
Al final de la creación de una nueva generación, mutamos los individuos. La mutación de cada uno consiste en elegir aleatoriamente, con una proporción del 0.02%, segundos y sobre cada uno de estos elegidos sumar o restar, con probabilidad, un entero entre 0 y el valor anterior por 1.5. De esta forma, el crecimiento será exponencial respecto de las generaciones. Como antes, la probabilidad crece exponencialmente respecto de los segundos y decrece sobre el tipo de mina, es decir, los últimos segundos cambiarán con mayor probabilidad la cantidad de minas compradas, siendo aún más probable este cambio para las minas baratas.  

### Cruce
Cruzamos cada individuo con su siguiente en la población. El cruce devuelve dos nuevos individuos. Se realiza sobre cada segundo tomando el primer tercio y el último de un padre y el restante del otro. El padre con mayor cantidad de genes se elige aleatoriamente.

En base a un parámetro (proporción de cruces), una parte de la población pasa a la siguiente generación sin cruzarse.

## Función de evaluación
La función de evaluación consiste simplemente en la ejecucción del juego comprando minas según indica el cromosoma en sus genes. Cada vez que se realice una compra inválida ese individuo tendrá automáticamente un fitness de -1.

## Parámetros
- Número de torneo: 3 individuos en los grupos del torneo.
- Número de generaciones: 200
- Proporción de cruces: 80%
- Tamaño de la población: 20 (un valor mayor solo reduce el rendimiento)

## Resultados
Como resultados finales hemos utilizado el siguiente script para realizar 100 ejecucciones y tomar la media de ellas. Mostramos el número de dígitos de la solución porque al ser números tan grandes lo consideramos más representativo. Las ejecucciones se realizan en paralelo/concurrente para que sean más rápidas.

In [1]:
%%time
import tests

tests.main()

Final: 18.135306722319243
Final: 18.12011546544637
Final: 18.15787333972489Final: 18.17122346319397

Final: 18.10337701314912
Final: 18.181534870957087
Final: 18.137602722575338
Final: 18.199335090600076
Final: 18.22052615383605
Final: 18.18973658953762
Final: 18.157160864770585
Final: 18.13911680626744
Final: 18.167188285726347
Final: 18.065872382242006
Final: 18.134050807928976
Final: 18.171495541852526
Final: 18.15381939533121
Final: 18.154242891951295
Final: 18.11234480867173
Final: 18.129356054467024
Final: 18.123875907364933
Final: 18.123464309167225
Final: 18.113139981819312
Final: 18.142633506075004
Final: 18.130029761860676
Final: 18.11053315823033
Final: 18.194957978108125
Final: 18.104912547231056
Final: 18.06761850091883
Final: 18.159883288355555
Final: 18.14827374662919
Final: 18.09403088135662
Final: 18.12723772143739
Final: 18.08086851274746
Final: 18.112153214228506
Final: 18.135401232242604
Final: 18.194259777815454
Final: 18.127743303712776
Final: 18.155194310657404
F

El resultado final que hemos obtenido ha sido de **18.16 dígitos** en 4min 40s, pero el tiempo depende enormemente del ordenador (en concreto, del número de núcleos del procesador).